### PR TITLE
Issue 1601: Bugfix for error handling

### LINKED ISSUE #1601: StateSynchronizerImpl  continues to try to connect to old SSS post SSS failover
**Problem description**
StateSynchronizerImpl continues to connect to old SSS even though SSS has failed over and old SSS no longer exists. 
This results in client getting CFE. 

**Problem location**
StateSynchronizerImpl
`io.pravega.common.util.RetriesExhaustedException: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException: io.netty.channel.AbstractChannel$AnnotatedConnectException: syscall:getsockopt(...): /<OldIp>:12345
        at io.pravega.common.util.Retry$RetryAndThrowBase.run(Retry.java:207)
        at io.pravega.client.segment.impl.SegmentMetadataClientImpl.fetchCurrentStreamLength(SegmentMetadataClientImpl.java:243)
        at io.pravega.client.state.impl.RevisionedStreamClientImpl.readFrom(RevisionedStreamClientImpl.java:102)
        at io.pravega.client.state.impl.StateSynchronizerImpl.fetchUpdates(StateSynchronizerImpl.java:75)
        at io.pravega.client.stream.impl.ReaderGroupStateManager.fetchUpdatesIfNeeded(ReaderGroupStateManager.java:261)
        at io.pravega.client.stream.impl.ReaderGroupStateManager.acquireNewSegmentsIfNeeded(ReaderGroupStateManager.java:251)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.acquireSegmentsIfNeeded(EventStreamReaderImpl.java:170)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.updateGroupStateIfNeeded(EventStreamReaderImpl.java:143)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEvent(EventStreamReaderImpl.java:82)
        at io.pravega.controller.eventProcessor.impl.EventProcessorCell$Delegate.run(EventProcessorCell.java:98)
        at com.google.common.util.concurrent.AbstractExecutionThreadService$1$2.run(AbstractExecutionThreadService.java:66)
        at com.google.common.util.concurrent.Callables$4.run(Callables.java:122)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.util.concurrent.CompletionException: io.pravega.shared.protocol.netty.ConnectionFailedException: io.netty.channel.AbstractChannel$AnnotatedConnectException: syscall:getsockopt(...): /<OldIP>:12345
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292)
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308)
        at java.util.concurrent.CompletableFuture.uniAccept(CompletableFuture.java:647)
        at java.util.concurrent.CompletableFuture.uniAcceptStage(CompletableFuture.java:669)
        at java.util.concurrent.CompletableFuture.thenAccept(CompletableFuture.java:1997)
        at io.pravega.client.segment.impl.SegmentMetadataClientImpl.getSegmentInfo(SegmentMetadataClientImpl.java:187)
        at io.pravega.client.segment.impl.SegmentMetadataClientImpl.lambda$fetchCurrentStreamLength$7(SegmentMetadataClientImpl.java:244)
        at io.pravega.common.util.Retry$RetryAndThrowBase.run(Retry.java:190)
        ... 12 common frames omitted
Caused by: io.pravega.shared.protocol.netty.ConnectionFailedException: io.netty.channel.AbstractChannel$AnnotatedConnectException: syscall:getsockopt(...): /172.16.24.3:12345
        at io.pravega.client.netty.impl.ConnectionFactoryImpl$2.operationComplete(ConnectionFactoryImpl.java:125)
        at io.pravega.client.netty.impl.ConnectionFactoryImpl$2.operationComplete(ConnectionFactoryImpl.java:119)
        at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:507)
        at io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:500)
        at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:479)
        at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:420)
        at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:122)
        at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.fulfillConnectPromise(AbstractEpollStreamChannel.java:910)
        at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.finishConnect(AbstractEpollStreamChannel.java:929)
        at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollOutReady(AbstractEpollStreamChannel.java:946)
        at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:384)
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:299)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
        at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:144)
        ... 1 common frames omitted
Caused by: io.netty.channel.AbstractChannel$AnnotatedConnectException: syscall:getsockopt(...): /172.16.24.3:12345
`

**Suggestions for an improvement**
Handle all CFE scenarios and ensure segment address is refreshed from controller in case of a connection failure.

### PR DESCRIPTION
**Pull request by tkaitchuck**
**Change log description**
Upon error clear the connection before re-connecting so that any retries have the opportunity to go to a different host.

**Purpose of the change**
Fixes #1601. There was a case where ConnectionFailedException was being set on the future for the connection establishment, but the connection was not replaced when getConnection() was retried.

**What the code does**
Rather than only catching known exceptions and assuming that exceptions are only wrapped once, it makes sure that all possible exception paths result in connectionClose being called. 

**How to verify it**
Run the System tests.

### CODE REVIEW

File: client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java [hunk 1]
```diff
@@ -421,7 +421,12 @@ void setupConnection() throws ConnectionFailedException {
             ClientConnection connection = getAndHandleExceptions(newConnection, ConnectionFailedException::new);
             state.newConnection(connection);
             SetupAppend cmd = new SetupAppend(requestIdGenerator.get(), writerId, segmentName);
-            connection.send(cmd);
+            try {
+                connection.send(cmd);
+            } catch (ConnectionFailedException e) {
```
Comment (shiveshr) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java]: shouldnt we have throwable here too?

File: client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java [hunk 2]
```diff
@@ -454,7 +459,12 @@ public void flush() throws SegmentSealedException {
                          .throwingOn(SegmentSealedException.class)
                          .run(() -> {
                              ClientConnection connection = getConnection();
-                             connection.send(new KeepAlive());
+                             try {
+                                 connection.send(new KeepAlive());
+                             } catch (ConnectionFailedException e) {
```
Comment (shiveshr) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java]: same as above

File: common/src/main/java/io/pravega/common/ExceptionHelpers.java [hunk 3]
```diff
@@ -60,8 +60,8 @@ public static boolean shouldUnwrap(Class<? extends Exception> c) {
      * @return The cause or the exception provided.
      */
     public static Throwable unwrapIfRequired(Throwable e) {
-        if ((e instanceof CompletionException || e instanceof ExecutionException) && e.getCause() != null) {
-            return e.getCause();
+        while ((e instanceof CompletionException || e instanceof ExecutionException) && e.getCause() != null) {
```
Comment (shiveshr) [inline common/src/main/java/io/pravega/common/ExceptionHelpers.java]: Orthogonal comment: this method seems similar to getRealException with only difference being IO exception. 
do we need two methods?

Comment (tkaitchuck) [inline common/src/main/java/io/pravega/common/ExceptionHelpers.java]: Yes. I did some digging and I think these can be combined.

File: client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java (hunk 1 above)
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java]: Changed to Exception

File: client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java (hunk 2 above)
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java]: Changed to Exception

**Pull request merged**